code/messaging/src/phone/java/com/adobe/marketing/mobile/Messaging.java

```diff
@@ -191,20 +204,116 @@ public static void refreshInAppMessages() {
+                        final Surface surface = Surface.fromUriString(DataReader.optString(propositionData, SCOPE, null));
+                        final List<Proposition> propositions = new ArrayList<>();
+                        final Proposition proposition = Proposition.fromEventData(propositionData);
+                        propositions.add(proposition);
+                        convertedPropositions.put(surface, propositions);
+                    }
+                }
+                propositionsResponseHandler.call(convertedPropositions);
+            });
+        } else {
+            isPropositionsResponseListenerRegistered = false;
+        }
     }
 
-    // TODO: implement
+    /**
+     * Dispatches an event to retrieve the previously fetched (and cached) feeds content from the SDK for the provided surfaces.
+     * If the feeds content for one or more surfaces isn't previously cached in the SDK, it will not be retrieved from Adobe Journey Optimizer via the Experience Edge network.
+     *
+     * @param surfaces A {@link List<Surface>} containing {@link Surface}s to be used for retrieving previously fetched propositions
+     * @param callback A {@link AdobeCallback} which will be invoked with a {@link Map<Surface, List<Proposition>>} containing previously fetched feeds content
+     */
     public static void getPropositionsForSurfaces(@NonNull final List<Surface> surfaces, @NonNull final AdobeCallback<Map<Surface, List<Proposition>>> callback) {
+        if (callback == null ) {
+            Log.warning(LOG_TAG, CLASS_NAME, "Cannot get propositions as the provided callback is null.");
+            return;
+        }
+
+        if (surfaces == null || surfaces.isEmpty()) {
+            Log.warning(LOG_TAG, CLASS_NAME, "Cannot get propositions as the provided list of surfaces is null or empty.");
+            return;
+        }
+
+        final List<String> validSurfacePaths = new ArrayList<>();
+        for (final Surface surface : surfaces) {
+            if (surface.isValid()) {
+                validSurfacePaths.add(surface.getUri());
+            }
+        }
+
+        if (validSurfacePaths.isEmpty()) {
+            Log.warning(LOG_TAG, CLASS_NAME, "Cannot get propositions as the provided list of surfaces has no valid items.");
+            return;
+        }
 
+        final Map<String, Object> eventData = new HashMap<>();
+        eventData.put(GET_PROPOSITIONS_EVENT, true);
+        eventData.put(SURFACES, validSurfacePaths);
+
+        final Event getPropositionsEvent = new Event.Builder(GET_PROPOSITIONS,
+                EventType.MESSAGING, EventSource.REQUEST_CONTENT)
+                .setEventData(eventData)
+                .build();
+
+        MobileCore.dispatchEventWithResponseCallback(getPropositionsEvent, 2000, new AdobeCallbackWithError<Event>() {
+            @Override
+            public void fail(final AdobeError adobeError) {
+                callback.call(null);
```

Comment:
If provided callback is an instance of `AdobeCallbackWithError` then you need to invoke `fail` method on it here. See https://github.com/adobe/aepsdk-optimize-android/blob/ajo-inbound-mobile/code/optimize/src/main/java/com/adobe/marketing/mobile/optimize/Optimize.java#L382